content/gotraining.slide

```diff
@@ -212,21 +224,265 @@ Execute with
+
+
+* Recommendations for CD @ANZ (1/2)
+
+- Use *trunk* *based* *development* (abandon GitFlow)
+- Create *small* *PRs* (Pull Requests) that can be Squash Merged
+- ... which lead to revertible commits that have passed automated tests
+- Protect `master` branch on GitHub
+- Write great commit messages
+- Think about the reviewer and review your own PRs first
+
+
+* Recommendations for CD @ANZ (2/2)
+
+- *Lint* strictly `golangci-lint` `run`
+- Ensure *100%* *coverage*
+- Test (unit & system), lint & cover as part of *CI* (Continuous Integration)
+- *Version* all deployments (use [[https://semver.org/][semver]]; tag source, images etc.)
+- *Document* for newbies in the README.md (must!) and use godoc
+- ... Non-prod vs Prod ?
+- Enjoy your CD as last step after successful CI run on master
+
+
+* Protect `master` branch on GitHub
+
+	Settings -> Options:
+	[ ] Allow merge commits
+	[x] Allow squash merging
+	[ ] Allow rebase merging
+
+	Settings -> Branches -> [Add rule|Edit] `master`:
+	[x] Require pull request reviews before merging
+	[x] Dismiss stale pull request approvals when new commits are pushed
+	[x] Require status checks to pass before merging
+	[x] Require branches to be up to date before merging
+	[x] Include administrators
+
+
+* Commit messages
+
+.image img/xkcd_git_commit.png
+.caption [[https://xkcd.com/1296/]]
+
+
+* Commit messages - spot the difference
+
+[[https://chris.beams.io/posts/git-commit/]]
+
+	$ git log --oneline -5 --author cbeams --before "Fri Mar 26 2009"
+
+	e5f4b49 Re-adding ConfigurationPostProcessorTests after its brief removal in r814. @Ignore-ing the testCglibClassesAreLoadedJustInTimeForEnhancement() method as it turns out this was one of the culprits in the recent build breakage. The classloader hacking causes subtle downstream effects, breaking unrelated tests. The test method is still useful, but should only be run on a manual basis to ensure CGLIB is not prematurely classloaded, and should not be run as part of the automated build.
+	2db0f12 fixed two build-breaking issues: + reverted ClassMetadataReadingVisitor to revision 794 + eliminated ConfigurationPostProcessorTests until further investigation determines why it causes downstream tests to fail (such as the seemingly unrelated ClassPathXmlApplicationContextTests)
+	147709f Tweaks to package-info.java files
+	22b25e0 Consolidated Util and MutableAnnotationUtils classes into existing AsmUtils
+	7f96f57 polishing
+
+versus
+
+	$ git log --oneline -5 --author pwebb --before "Sat Aug 30 2014"
+
+	5ba3db6 Fix failing CompositePropertySourceTests
+	84564a0 Rework @PropertySource early parsing logic
+	e142fd1 Add tests for ImportSelector meta-data
+	887815f Update docbook dependency and generate epub
+	ac8326d Polish mockito usage
+
+
+* Recipe for great commit messages
+
+[[https://chris.beams.io/posts/git-commit/]]
+
```

Comment:
it might also be worth introducing `git add -p` e.g. when you make fixes for two different defects while editing the one file at the same time, and want to commit them separately

Reply:
Done.